### PR TITLE
WIP: AutoSchema: Inherit DRF AutoSchema

### PR DESCRIPTION
**Pull request by jayvdb**
Replace custom _map_basic_serializer with _map_field,
and invoke super _map_serializer as a fallback.

DRF AutoSchema PR 7257 makes all of these fields public,
so Spectacular AutoSchema should implement them and use
them when appropriate for interoperability.

Related to https://github.com/tfranzel/drf-spectacular/issues/31
Related to https://github.com/tfranzel/drf-spectacular/issues/45

### CODE REVIEW

File: drf_spectacular/openapi.py [hunk 1]
```diff
@@ -541,55 +541,28 @@ def _map_min_max(self, field, content):
...
+            if hasattr(DRFAutoSchema, 'map_serializer'):
+                result = super().map_serializer(serializer)
+            else:
+                result = super()._map_serializer(serializer)
 
-            schema = self._map_serializer_field(field)
+                if result.get('properties'):
+                    # Move 'type' to top
+                    new = {'type': 'object'}
+                    new.update(result)
```
Comment (jayvdb) [inline drf_spectacular/openapi.py:559]: this is only needed to minimise changes to the .yml file.

File: drf_spectacular/openapi.py [hunk 2]
```diff
@@ -624,6 +597,33 @@ def _map_field_validators(self, field, schema):
...
+        result.update(schema)
+        if result.get('properties'):
+            result['type'] = 'object'
+
+        # sibling entries to $ref will be ignored as it replaces itself and its context with
+        # the referenced object. Wrap it in a separate context.
+        if '$ref' in result and len(result) > 1:
+            return {'allOf': [{'$ref': schema.pop('$ref')}], **schema}
+
+        new = {}
```
Comment (jayvdb) [inline drf_spectacular/openapi.py:613]: this is only needed to minimise changes to the .yml file.

File: tests/test_fields.yml [hunk 3]
```diff
@@ -161,8 +161,10 @@ components:
           minimum: -1000
         field_file:
           type: string
+          format: binary
```
Comment (jayvdb) [inline tests/test_fields.yml:164]: This is a byproduct of using DRF's `map_field`.  I can avoid it if it is problematic.

**Comment by jayvdb** [comment]
Currently fails on drf3.10 and fails test_recursion

**Comment by tfranzel** [comment]
hi! could you comment on the motivation for this change? was something broken, because now it is. we can talk about slight schema changes, but that change raises exceptions and has recursion loops.

that inheritance choice was actually on purpose. interoperability is broken on several places as those two implementation diverged significantly. 

also i fixed the travis integration and reran all PR. apparently sometimes those oauth2 credentials break without anything changing.

**Comment by jayvdb** [comment]
https://github.com/openwisp/django-rest-framework-gis/pull/223/files#r418517749 gives a bit of the motivation.

> was something broken, because now it is. .. but that change raises exceptions and has recursion loops.

It is a WIP.  As mentioned above "Currently fails on drf3.10 and fails test_recursion".

I'll attempt to the recursion with proper inheritance, but it might require a dirty hack.

drf3.10 is just time and effort that isnt a priority for me ATM while there are other larger problems to solve.

IMO yes, it is "broken" to use DRF's `DEFAULT_SCHEMA_CLASS` and not implement it as a class which is not compatible even at the most basic level.

**Comment by tfranzel** [comment]
ahh now i understand! ... and that is exactly the reason why DRF's recommended usage of AutoSchema just does not scale. I read in several DRF PR that people should just subclass the AutoSchema to their needs. Problem is that if everybody does that, you get 4 incompatible AutoSchemas for 4 libraries you are using. sounds very brittle to me.

i'm afraid your approach also might not scale that well. if i understand correctly, you want to fuse those 2 implementations together via "mixin" inheritance and make spectacular aware of that. the problem i see there is that implementations might break unexpected things in their subclass. that looks very tricky.

all i can think of is inventing a kind of AutoSchema adapter for 3rd party lib AutoSchema implementations, so basically [composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance). That would keep multiple implementations separate from each other and other 3rd party libs. 

what do you think? and btw how is that django-oscar-api coming?

File: drf_spectacular/openapi.py [hunk 4]
```diff
@@ -541,58 +542,44 @@ def _map_min_max(self, field, content):
         if field.min_value:
             content['minimum'] = field.min_value
 
-    def _map_serializer(self, serializer, direction):
+    def _map_serializer(self, serializer, direction=None):
         serializer = force_instance(serializer)
         serializer_extension = OpenApiSerializerExtension.get_match(serializer)
 
-        if serializer_extension:
+        if serializer_extension and direction:
```
Comment (jayvdb) [inline drf_spectacular/openapi.py:549]: This ` and direction` is optional, and probably needs to be removed.

**Comment by jayvdb** [comment]
All green.  The horrible hack for recursion error isnt permanent - it only isolates that problem until I get to solving it properly.

> Problem is that if everybody does that, you get 4 incompatible AutoSchemas for 4 libraries you are using. sounds very brittle to me.

Sure it *can* be brittle *if*(when) they are incompatible, but that is assuming the worst case, and can be solved by Django apps working together, which is done all the time, and is increasing as the Django ecosystem becomes more stable with larger and more inclusive maintainer teams replacing single-developer-maintainer who cant scale/collaborate/endure/etc.  Incompatibilities are constantly 'fixed', re-broken, 'fixed' again, etc.  And hopefully integration tests added on one or both sides when someone is motivated enough to prevent future breakages.

> you want to fuse those 2 implementations together via "mixin" inheritance

Right.  Multiple inheritance. The PR as-is doesnt quite achieve it wrt drf-gis, but getting close.

**Comment by jayvdb** [comment]
I've got it working with https://github.com/jayvdb/django-rest-framework-gis/commit/85983bc0fbd7b811be2a95c53bf23223e8cabe9b ; but did have a few more drf-spectacular changes needed which I'll add to this PR tmr.

Worth noting that this approach for GIS currently results in an ugly schema because the DRF-GIS schema doesnt use `components` and `$ref` - I could enhance drf-spectacular to auto-promote "large" field schemas to a `component`, but that is probably going a step too far for me, so I am going to go back to https://github.com/tfranzel/drf-spectacular/pull/38 .

I still think this PR is useful and important, but it isnt the best path for GIS at the moment.

File: tests/test_fields.yml (hunk 3 above)
Comment (jayvdb) [reply tests/test_fields.yml:164]: Diagnosed and raised as https://github.com/tfranzel/drf-spectacular/pull/69

**Comment by tfranzel** [comment]
stale and not the direction we want to go in atm

**Pull request closed without merging**